### ISSUE TITLE
ReferenceType has undefined type

### ISSUE THREAD

**Issue by awinograd** [vega/ts-json-schema-generator#45]
I get the an undefined error trying to parse `MyType`.

```
import { View, ViewStyle, StyleProp } from 'react-native';
export type MyType = StyleProp<ViewStyle>
```

```
TypeError: Cannot read property 'getId' of undefined
    at ReferenceType.getId (/Users/awinograd/programming/ts-json-schema-generator/dist/src/Type/ReferenceType.js:12:26)                                                                                    
    at (.types.map (/Users/awinograd/programming/ts-json-schema-generator/dist/src/Type/UnionType.js:11:25)                                                                                                
    at Array.map (<anonymous>)
    at UnionType.getId (/Users/awinograd/programming/ts-json-schema-generator/dist/src/Type/UnionType.js:10:33)                                                                                            
    at ids.join.<.context.getArguments.map (/Users/awinograd/programming/ts-json-schema-generator/dist/src/CircularReferenceNodeParser.js:33:78)                                                           
    at Array.map (<anonymous>)
    at CircularReferenceNodeParser.createCacheKey (/Users/awinograd/programming/ts-json-schema-generator/dist/src/CircularReferenceNodeParser.js:33:61)                                                    
    at CircularReferenceNodeParser.createType (/Users/awinograd/programming/ts-json-schema-generator/dist/src/CircularReferenceNodeParser.js:13:26)                                                        
    at ChainNodeParser.createType (/Users/awinograd/programming/ts-json-schema-generator/dist/src/ChainNodeParser.js:17:50)                                                                                
    at ExpressionWithTypeArgumentsNodeParser.createType (/Users/awinograd/programming/ts-json-schema-generator/dist/src/NodeParser/ExpressionWithTypeArgumentsNodeParser.js:23:41) 
```

So far, I've figured out that `getId` is called in `createType` before the returned value is actually set on the reference on the same line in `CircularReferenceNodeParser`

https://github.com/vega/ts-json-schema-generator/blob/c0f13a0c8aa01a4bba22c0e46de33585677a39b0/src/CircularReferenceNodeParser.ts#L26

**Comment by domoritz** [comment]
I recommend https://astexplorer.net/ to look at what is actually defined.

**Comment by awinograd** [comment]
Thanks, that's a helpful resource!

**Comment by awinograd** [comment]
Here's a minimal repro in astexplorer.net for reference

https://astexplorer.net/#/gist/f418932fb2afef0eb5a16cd9fb2598f1/79e310c9b62ea6ef55e558242cb530c810743c68

**Comment by RegisImbellus** [comment]
Hit a similar error:
```
.../node_modules/ts-json-schema-generator/dist/ts-json-schema-generator.js:32
        throw error;
        ^

TypeError: Cannot read property 'getId' of undefined
    at ReferenceType.getId (.../node_modules/ts-json-schema-generator/dist/src/Type/ReferenceType.js:6:26)
    at ids.join.<.context.getArguments.map (.../node_modules/ts-json-schema-generator/dist/src/CircularReferenceNodeParser.js:29:78)
    at Array.map (<anonymous>)
    at CircularReferenceNodeParser.createCacheKey (.../node_modules/ts-json-schema-generator/dist/src/CircularReferenceNodeParser.js:29:61)
    at CircularReferenceNodeParser.createType (.../node_modules/ts-json-schema-generator/dist/src/CircularReferenceNodeParser.js:13:26)
    at ChainNodeParser.createType (.../node_modules/ts-json-schema-generator/dist/src/ChainNodeParser.js:17:50)
    at ExpressionWithTypeArgumentsNodeParser.createType (.../node_modules/ts-json-schema-generator/dist/src/NodeParser/ExpressionWithTypeArgumentsNodeParser.js:23:41)
    at ChainNodeParser.createType (.../node_modules/ts-json-schema-generator/dist/src/ChainNodeParser.js:17:50)
    at AnnotatedNodeParser.createType (.../node_modules/ts-json-schema-generator/dist/src/NodeParser/AnnotatedNodeParser.js:15:47)
    at baseType.types.map (.../node_modules/ts-json-schema-generator/dist/src/NodeParser/InterfaceNodeParser.js:33:72)
```

Triggered by the following typescript:
```
export type AnyJson = boolean | number | string | null | JsonArray | JsonMap;
export interface JsonMap { [key: string]: AnyJson; }
export interface JsonArray extends Array<AnyJson> {}
```

Removing `JsonArray` from AnyJson union fixes the issue and allows the schema to generate.

**Comment by kayahr** [comment]
Another simple way to reproduce the error:

```typescript
type Map<T> = { [ key: string]: T; };

export type Test = {
    a: Map<Test>;
};
```

For some reason it works when inlining the Map type:

```typescript
export type Test = {
    a: { [ key: string]: Test; };
};
```

**Comment by kayahr** [comment]
I think I found the problem:

The class `CircularReferenceNodeParser` is obviously responsible for solving circular reference problems but in this case it doesn't work. This parser creates an empty ReferenceType which is not yet connected to the real type and stores this type in a map called `circular`. Other types which are created while creating the actual type will get this unfinished ReferenceType instead of producing an endless loop. This is fine but the problem is that for TypeReferenceNodes this unfinished ReferenceType is pushed into the arguments array on the context. And from these argument types the IDs are read when creating a key for the circular map. But this ID can not be read because the ReferenceType is not really referencing a type yet. And that's why it fails...

Or in other words this happens (with the example from my previous comment):

1. Starting creation of type for TypeAliasDeclaration `export type Test = { a: Map<Test>; }`
2. This starts the creation of type for TypeLiteral `{ a: Map<Test>; }`
3. This starts the creation of type for TypeReference `Map<Test>`
4. This starts the creation of type for TypeReference `Test`
5. This again starts the creation of type for TypeAliasDeclaration `export type Test = { a: Map<Test>; }` which now returns the unfinished ReferenceType from step 1.
6. This unfinished reference type is pushed to the arguments array of the context.
7. createType is called again, this time for the TypeAliasDeclaration `type Map<T> = { [ key: string]: T; }`
8. Key generation for this new ReferenceType now fails because this calls `getId()` on the unfinished ReferenceType from step 1 which doesn't have a referenced type yet.

Unfortunately I don't know yet how to solve this...

**Comment by kayahr** [comment]
Maybe it would be possible to separate the Type-ID creation from the actual Type creation so the ID is known to the ReferenceType before the actual type is created and therefor before the creation of other types are triggered by it?

To check if this would be possible I created an ugly workaround:

https://github.com/vega/ts-json-schema-generator/compare/master...iplabs:workaround/circular-type-id?expand=1

With this workaround the ReferenceType can now access the type ID via a WeakMap using the `ts.Node` as key. I changed the `InterfaceNodeParser` and the `TypeAliasNodeParser` to put the generated ID into this map before creating the actual type, so with this workaround all the examples above are now working.

As already said, I only consider this an ugly workaround. There must be a cleaner way to solve this.

**Comment by kayahr** [comment]
Ah, forget the ugly workaround with the WeakMap. I guess it may not work always anyway because the ts.Node without the context is a bad WeakMap key.

But I think I found a better solution: I added an optional ReferenceType argument to to `NodeParser::createType()` method. The CircularReferenceNodeParser passes the created ReferenceType via this argument to the actual node parser and each node parser can then decide if it wants to set the ID and name (Yes, that's also necessary, because ExposeNodeParser accesses it before type is finished) on the reference node before the actual type is created.

Still a little bit hacky but PR is incoming anyway.
